proposals/0292-package-registry-service.md

```diff
@@ -122,81 +273,35 @@ There are currently two concrete implementations of `PackageContainer`:
-only when the user passes the `--enable-package-registries` command-line flag.
-This may change in a future release.
-
-```terminal
-$ swift build --enable-package-registries
-```
-
-When package registries are enabled,
-Swift Package Manager will first attempt to use package registry API calls
-to resolve qualifying dependencies,
-falling back to Git operations if those API calls fail.
-A dependency qualifies for resolution through a package registry
-if it satisfies all of the following criteria:
-
-* The package has a url with an `https` scheme.
-* The last path component of the package url has no file extension.
-* The dependency specifies an exact version or range of versions.
-
-For example,
-here are a list of dependencies that do and do not qualify:
-
-```swift
-// ✅ These dependencies qualify for resolution with package registry
-.package(url: "https://github.com/mona/LinkedList", from: "1.1.0")
-.package(url: "https://github.com/mona/LinkedList", .exact("1.1.0"))
-.package(url: "https://github.com/mona/LinkedList", .upToNextMajor(from: "1.1.0"))
-.package(url: "https://github.com/mona/LinkedList", .upToNextMinor(from: "1.1.0"))
-
-// ❌ These dependencies can only be resolved using Git
-.package(url: "git@github.com:mona/LinkedList.git", from: "1.1.0") // No https scheme
-.package(url: "https://github.com/mona/LinkedList.git", from: "1.1.0") // .git file extension
-.package(url: "https://github.com/mona/LinkedList", .branch("master")) // No version
-.package(url: "https://github.com/mona/LinkedList", .revision("d6ca4e56219a8a5f0237d6dcdd8b975ec7e24c89")) // No version
-.package(path: "../LinkedList") // No https scheme or version
-```
+| Task                                  | Git operation               | Registry request                                     |
+| ------------------------------------- | --------------------------- | ---------------------------------------------------- |
+| Fetch the contents of a package       | `git clone && git checkout` | `GET /{scope}/{name}/{version}.zip`           |
+| List the available tags for a package | `git tag`                   | `GET /{scope}/{name}`                         |
+| Fetch a package manifest              | `git clone`                 | `GET /{scope}/{name}/{version}/Package.swift` |
 
 Package registries support
 [version-specific _manifest_ selection][version-specific-manifest-selection]
 by providing a list of versioned manifest files for a package
 (for example, `Package@swift-5.3.swift`)
-in its response to `GET /{package}/{version}/Package.swift`.
-However, package registries won't support
-[version-specific _tag_ selection][version-specific-tag-selection],
-and instead rely on [Semantic Versioning][SemVer]
-to accomodate different versions of Swift
-(for example,
-by using major release versions
-or build metadata like `1.0.0+swift-5_3`).
+in its response to `GET /{scope}/{name}/{version}/Package.swift`.
```

Comment:
Isn't the response to `GET /{scope}/{name}/{version}/Package.swift` the contents of the primary `Package.swift` file? How does the registry provide the list of available versioned manifest files?

This seems similar to [my comment/question](https://github.com/mattt/swift-evolution/commit/07497a799f80ecb80bea163b84e6f31cc331bc91#commitcomment-48331998) of how the registry would provide the list of available distributions.

Should we consider having the response returned by  `GET /{scope}/{name}` include not just a list of releases but a map instead, where each version is a key and the value is an object that includes the available files for that release (the Package.swift, other version-specific manifest files, the .zip and other distributions such as .xcframework).

Alternatively, instead having this information returned for all versions in the `GET /{scope}/{name}` response, what about replacing the `GET /{scope}/{name}/Package.swift` request with something slightly different that includes all variations of the Package.swift and links to all distributions (.zip, .xcframework)?

Reply:
The list of available manifests is provided by the response to `GET /{scope}/{name}/{version}/Package.swift`. From [4.3. Fetch manifest for a package release](https://github.com/apple/swift-package-manager/blob/e593c129d6438c509b4220321cbd26b804e7733e/Documentation/Registry.md#43-fetch-manifest-for-a-package-release):

> A server SHOULD include Link header fields with the alternate relation type for each additional file in the release's source archive whose filename matches the following regular expression pattern:
>
> ```regexp
> \APackage(?:@swift-\d+(?:\.\d+){0,2})?.swift\z
> ```

Reply:
So to be clear, if a client wants to use a versioned manifest it will end up making 4 HTTP requests instead of 3:
 1. `GET /{scope}/{name}`
 2. `GET /{scope}/{name}/{version}/Package.swift`
 3. `GET /{scope}/{name}/{version}/Package.swift{swift-version}` (this is the extra request)
 4. `GET /{scope}/{name}/{version}.zip`

That is one of the reasons why I was suggesting alternatives that would return the information in a variation of the first or second request (e.g. the first request could return a list of all available "files").

Reply:
You're correct that Swift Package Manager would be make one or more `GET /{scope}/{name}/{version}/Package.swift{swift-version}` requests to use a versioned manifest file. However, that original request to the unqualified `Package.swift` shouldn't be considered extraneous. The [version-specific manifest selection process](https://github.com/apple/swift-package-manager/blob/main/Documentation/Usage.md#version-specific-manifest-selection) needs that unversioned manifest to make a determination. Knowledge of the which version-specific manifests are available ahead of time wouldn't eliminate the need to fetch `Package.swift` in the first place.

Reply:
> Knowledge of the which version-specific manifests are available ahead of time wouldn't eliminate the need to fetch Package.swift in the first place.

Thanks for the clarification. I'm less familiar with Swift Package Manager than others. I can say that most of the protocols that I'm better familiar with, such as npm, pypi, nuget, only require the client to make 2 requests per package dependency (Maven is an outlier in the number of requests).

Reply:
@mattt @abertelrud @yim-lee @neonichu @mmarston would it be better if instead of `GET /{scope}/{name}/{version}/Package.swift` we do `GET /{scope}/{name}/{version}/manifests` and return all the manifests in one go?